pkg/ottl/context_inferrer.go

```diff
@@ -0,0 +1,85 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottl // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+
+import "math"
+
+var (
+	defaultContextInferPriority = []string{
+		"log",
+		"metric",
+		"datapoint",
+		"spanevent",
+		"span",
+		"resource",
+		"scope",
+		"instrumentation_scope",
+	}
+)
+
+// ContextInferrer is an interface used to infer the OTTL context from statements paths.
+type ContextInferrer interface {
+	// Infer returns the OTTL context inferred from the given statements paths.
+	Infer(statements []string) (string, error)
+}
+
+type priorityContextInferrer struct {
+	contextPriority      map[string]int
+	ignoreUnknownContext bool
+}
+
+func (s *priorityContextInferrer) Infer(statements []string) (string, error) {
+	var inferredContext string
+	var inferredContextPriority int
+
+	for _, statement := range statements {
+		parsed, err := parseStatement(statement)
+		if err != nil {
+			return inferredContext, err
+		}
+
+		for _, p := range getParsedStatementPaths(parsed) {
+			pathContextPriority, ok := s.contextPriority[p.Context]
+			if !ok {
+				if s.ignoreUnknownContext {
+					continue
+				}
+
+				// Lowest priority
+				pathContextPriority = math.MaxInt
+			}
+
+			if inferredContext == "" || pathContextPriority < inferredContextPriority {
+				inferredContext = p.Context
+				inferredContextPriority = pathContextPriority
+			}
+		}
+	}
+
+	return inferredContext, nil
+}
+
+// DefaultPriorityContextInferrer is like NewPriorityContextInferrer, but using the default
+// context priorities and ignoring unknown/non-prioritized contexts.
+func DefaultPriorityContextInferrer() ContextInferrer {
+	return NewPriorityContextInferrer(defaultContextInferPriority, false)
+}
+
+// NewPriorityContextInferrer creates a new priority-based context inferrer.
+// To infer the context, it compares all [ottl.Path.Context] values, prioritizing them based
+// on the provide contextsPriority argument, the lower the context position is in the array,
+// the more priority it will have over other items.
+// If unknown/non-prioritized contexts are found on the statements, they can be either ignored
+// or considered when no other prioritized context is found. To skip unknown contexts, the
+// ignoreUnknownContext argument must be set to false.
+func NewPriorityContextInferrer(contextsPriority []string, ignoreUnknownContext bool) ContextInferrer {
```

Comment:
My gut is saying that we'll come up with more options for this concept so I think it would be good to switch `ignoreUnknownContext bool` to the Options pattern if we keep it.